drivers/sdhc/sdhc_stm32.c

```diff
@@ -0,0 +1,777 @@
+
+#define STM32_SDHC_IRQ_HANDLER_FUNCTION(index)                                                     \
+	.irq_config_func = sdhc_stm32_irq_config_func_##index,
+
+#define STM32_SDHC_IRQ_HANDLER(index)                                                              \
+	static void sdhc_stm32_irq_config_func_##index(void)                                       \
+	{                                                                                          \
+		STM32_SDHC_IRQ_CONNECT_AND_ENABLE(index);                                          \
+	}
+
+#define SDHC_STM32_INIT(index)                                                                     \
+	STM32_SDHC_IRQ_HANDLER_DECL(index);                                                        \
+	static SDIO_HandleTypeDef hsd_##index = {                                                  \
+		.Instance = (MMC_TypeDef *)DT_INST_REG_ADDR(index),                                \
+	};                                                                                         \
+	static struct stm32_pclken pclken_##index[] = STM32_DT_INST_CLOCKS(index);                 \
+	PINCTRL_DT_INST_DEFINE(index);                                                             \
+	static const struct sdhc_stm32_config sdhc_stm32_cfg_##index = {                           \
+		.hsd = &hsd_##index,                                                               \
+		STM32_SDHC_IRQ_HANDLER_FUNCTION(index).pclken = pclken_##index,                    \
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(index),                                     \
+		.hw_flow_control = DT_INST_PROP(index, hw_flow_control),                           \
+		.clk_div = DT_INST_PROP(index, clk_div),                                           \
+		.bus_width = DT_INST_PROP(index, bus_width),                                       \
+		.power_delay_ms = DT_INST_PROP(index, power_delay_ms),                             \
+		.support_1_8_v = DT_INST_PROP(index, support_1_8_v),                               \
+		.sdhi_on_gpio = GPIO_DT_SPEC_GET_OR(DT_DRV_INST(index), sdhi_on_gpios, {0}),       \
+		.cd_gpio = GPIO_DT_SPEC_GET_OR(DT_DRV_INST(index), cd_gpios, {0}),                 \
+		.min_freq = DT_INST_PROP(index, min_bus_freq),                                     \
+		.max_freq = DT_INST_PROP(index, max_bus_freq),                                     \
+	};                                                                                         \
+	static struct sdhc_stm32_data sdhc_stm32_data_##index;                                     \
+	PM_DEVICE_DT_INST_DEFINE(index, sdhc_stm32_pm_action);                                     \
+	DEVICE_DT_INST_DEFINE(index, &sdhc_stm32_init, NULL, &sdhc_stm32_data_##index,             \
+			      &sdhc_stm32_cfg_##index, POST_KERNEL, CONFIG_SDHC_INIT_PRIORITY,     \
+			      &sdhc_stm32_api);                                                    \
+	STM32_SDHC_IRQ_HANDLER(index)
```

Comment:
It would help readability to have empty lines between the declarations/definitions:
```c
#define SDHC_STM32_INIT(index)									\
	STM32_SDHC_IRQ_HANDLER_DECL(index);							\
												\
	static SDIO_HandleTypeDef hsd_##index = {						\
		.Instance = (MMC_TypeDef *)DT_INST_REG_ADDR(index),				\
	};											\
												\
	static struct stm32_pclken pclken_##index[] = STM32_DT_INST_CLOCKS(index);		\
												\
	PINCTRL_DT_INST_DEFINE(index);								\
												\
	static const struct sdhc_stm32_config sdhc_stm32_cfg_##index = {			\
		.hsd = &hsd_##index,								\
		...
etc...